datalad/nonasyncrunner.py

```diff
@@ -0,0 +1,162 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command and function calls, allowing for dry runs and output handling
+
+"""
+
+import logging
+import os
+import queue
+import subprocess
+import threading
+import time
+from typing import Any
+
+
+logger = logging.getLogger("datalad.runner")
+
+STDOUT_FILENO = 1
+STDERR_FILENO = 2
+
+
+class ReaderThread(threading.Thread):
+    def __init__(self, file, a_queue, command):
+        super().__init__(daemon=True)
+        self.file = file
+        self.queue = a_queue
+        self.command = command
+        self.quit = False
+
+    def __str__(self):
+        return f"ReaderThread({self.file}, {self.queue}, {self.command})"
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the thread might be waiting in
+        os.read() or queue.put().
+        """
+        self.quit = True
```

Comment:
could we test this scenario/code block?

Reply:
The non blocking part is tested and we can wrap the blocking part into that (or do it seperately)

Reply:
I test the blocking part, although it is not a feature ;-)